### PR TITLE
Add Label smoothing in GANS.

### PR DESCRIPTION
**Pull request by walragatver**
This currently contains #1770 commits as well. 
I will add a small test for label smoothing soon.

### CODE REVIEW

**Comment by zoq** [comment]
Agreed a test would be great 👍

**Comment by walragatver** [comment]
Hi @zoq,
Ah, actually I have already added a test case. Please look at `gan_test.cpp`, in my last commit. Please let me know if that's incorrect.

**Comment by zoq** [comment]
Ahh, I missed the test, thanks!

File: src/mlpack/methods/ann/gan/gan.hpp [hunk 1]
```diff
@@ -95,16 +96,55 @@ class GAN
...
+   * Intialize the generator, discriminator and weights of the model for
+   * training.
+   * This function won't actually trigger training process.
+   *
+   * @param trainData The data points of real distribution.
+   * @param realLabel The label for data points from real distribution.
+   * @param fakeLabel The label for data points which are generated.
+   */
+  void ResetData(arma::mat trainData,
+                 double realLabel = 1.0,
```
Comment (zoq) [inline src/mlpack/methods/ann/gan/gan.hpp]: For consistency let's use ``const double`` for ``realLabel`` and ``fakeLabel``.

File: src/mlpack/methods/ann/gan/wgan_impl.hpp [hunk 2]
```diff
@@ -119,8 +133,10 @@ EvaluateWithGradient(const arma::mat& /* parameters */,
   generator.Forward(std::move(noise));
   predictors.cols(numFunctions, numFunctions + batchSize - 1) =
       boost::apply_visitor(outputParameterVisitor, generator.network.back());
-  responses.cols(numFunctions, numFunctions + batchSize - 1) =
-      -arma::ones(1, batchSize);
+
+  arma::mat fakeResponses(1, batchSize);
+  fakeResponses.fill(fakeLabel);
+  responses.cols(numFunctions, numFunctions + batchSize - 1) = fakeResponses;
```
Comment (zoq) [inline src/mlpack/methods/ann/gan/wgan_impl.hpp]: Correct me if I'm wrong but I guess:

```responses.cols(numFunctions, numFunctions + batchSize - 1).fill(fakeLabel)```

should work as well.

**Comment by ShikharJ** [comment]
@walragatver I would prefer if this PR was independent of the Serialization PR, as I see no dependency between label smoothing and serialization techniques. Also, it would be much easier to review that way.

File: src/mlpack/methods/ann/gan/gan.hpp [hunk 3]
```diff
@@ -95,16 +96,55 @@ class GAN
   //! Move constructor.
   GAN(GAN&&);
 
+  /**
+   * Intialize the generator, discriminator and weights of the model for
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/gan/gan.hpp]: This should be `Initialize`. Please make use of a spell-checker for US English.

File: src/mlpack/methods/ann/gan/gan.hpp [hunk 4]
```diff
@@ -95,16 +96,55 @@ class GAN
...
+   * training.
+   * This function won't actually trigger training process.
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/gan/gan.hpp]: Please continue the sentence on the same line.

**Comment by rcurtin** [comment]
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs.)

**Pull request closed without merging**